From d3727abfa79f74a5554051852e0e0cda31afeee9 Mon Sep 17 00:00:00 2001
From: Joseph Liu <kwliu@nuvoton.com>
Date: Tue, 24 Sep 2019 18:22:27 +0800
Subject: [PATCH] stmmac: Add eee fixup disable

Under some circumstances, a very high interrupt
rate is generated by lpi_irq on stmmac driver.
Workaround on some boards SoCs/boards can be
added to disable eee advertising.

Signed-off-by: Joseph Liu <kwliu@nuvoton.com>
---
 drivers/net/ethernet/stmicro/stmmac/stmmac_main.c     | 3 +++
 drivers/net/ethernet/stmicro/stmmac/stmmac_platform.c | 3 +++
 include/linux/stmmac.h                                | 1 +
 3 files changed, 7 insertions(+)

diff --git a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
index f826365c979d..37afe2ac94dd 100644
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -395,6 +395,9 @@ bool stmmac_eee_init(struct stmmac_priv *priv)
 	if (!priv->dma_cap.eee)
 		return false;
 
+	if (priv->plat->eee_force_disable)
+		return false;
+
 	mutex_lock(&priv->lock);
 
 	/* Check if it needs to be deactivated */
diff --git a/drivers/net/ethernet/stmicro/stmmac/stmmac_platform.c b/drivers/net/ethernet/stmicro/stmmac/stmmac_platform.c
index 170c3a052b14..c4873630c993 100644
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_platform.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_platform.c
@@ -465,6 +465,9 @@ stmmac_probe_config_dt(struct platform_device *pdev, const char **mac)
 	plat->en_tx_lpi_clockgating =
 		of_property_read_bool(np, "snps,en-tx-lpi-clockgating");
 
+	plat->eee_force_disable =
+		of_property_read_bool(np, "snps,eee-force-disable");
+
 	/* Set the maxmtu to a default of JUMBO_LEN in case the
 	 * parameter is not present in the device tree.
 	 */
diff --git a/include/linux/stmmac.h b/include/linux/stmmac.h
index dc60d03c4b60..3db5fe142817 100644
--- a/include/linux/stmmac.h
+++ b/include/linux/stmmac.h
@@ -164,6 +164,7 @@ struct plat_stmmacenet_data {
 	void (*exit)(struct platform_device *pdev, void *priv);
 	struct mac_device_info *(*setup)(void *priv);
 	void *bsp_priv;
+	int eee_force_disable;
 	struct clk *stmmac_clk;
 	struct clk *pclk;
 	struct clk *clk_ptp_ref;
-- 
2.17.1

